finality-aleph/src/session_map.rs

```diff
@@ -0,0 +1,528 @@
+use crate::{
+    first_block_of_session, session_id_from_block_num, ClientForAleph, SessionId, SessionPeriod,
+};
+use aleph_primitives::{AlephSessionApi, AuthorityId};
+use futures::StreamExt;
+use log::{debug, error, trace};
+use sc_client_api::{Backend, FinalityNotification};
+use sc_utils::mpsc::TracingUnboundedReceiver;
+use sp_runtime::{
+    generic::BlockId,
+    traits::{Block, Header, NumberFor},
+    SaturatedConversion,
+};
+use std::{collections::HashMap, marker::PhantomData, sync::Arc};
+use tokio::sync::{
+    oneshot::{Receiver as OneShotReceiver, Sender as OneShotSender},
+    Mutex,
+};
+
+type SessionMap = HashMap<SessionId, Vec<AuthorityId>>;
+type SessionSubscribers = HashMap<SessionId, Vec<OneShotSender<()>>>;
+
+pub trait AuthorityProvider<B> {
+    /// returns authorities for block `b`
+    fn authorities(&self, b: B) -> Option<Vec<AuthorityId>>;
```

Comment:
can we call the argument explicitly `block`?

Reply:
done